Move to a key identity in lieu of the config identifier hash. (#376)
tlswg/draft-ietf-tls-esni: draft-ietf-tls-esni.md

    struct {
        HpkeKdfId kdf_id;
        HpkeAeadId aead_id;

>     } HpkeSymmetricCipherSuite;

Reviewer: Why this name change?

Author: It seems to capture the intent of the struct better.

    struct {
        uint8 key_id;
        HpkeKemId kem_id;
        HpkePublicKey public_key;

>         HpkeSymmetricCipherSuite cipher_suites<4..2^16-4>;

Reviewer: IMO it's a bit tidier to have `cipher_suites` immediately precede the `kem_id`.

Author: I don't have a strong feeling here. I think the only important bit is that the kem_id precedes the public key.

Reviewer 2: Any change is a change and they are all equally not-free, so minimizing changes is not a goal I would worry about.  In terms of ordering, I prefer what @chris-wood proposed.

    } HpkeKeyConfig;

    struct {

>         HpkeKeyConfig key_config;

Reviewer 3: It's a little goofy to have the config_id apply to the whole thing here, but I suppose it's inlined.

Reviewer 3: I think we at least need a comment here to explain that.

Author: Can you suggest text?

The `ECHConfigContents` structure contains the following fields:

key_config
: A `HpkeKeyConfig` structure carrying the configuration information associated

> with the HPKE public key.

Reviewer 3: ```suggestion
with the HPKE public key. Note that this structure contains the `config_id` field, which applies to the
entire ECHConfigContents. Sites MUST NOT publish two differet`ECHConfigContents` values with the same
`HpkeKeyConfig` value.
```

maximum_name_length
: The longest name of a backend server, if known. If this value is not known it
can be set to zero, in which case clients SHOULD use the inner ClientHello
padding scheme described below. That could happen if wildcard names are in use,
or if names can be added or removed from the anonymity set during the lifetime
of a particular ECH configuration.

public_name
: The non-empty name of the client-facing server, i.e., the entity trusted to
update the ECH configuration. This is used to correct misconfigured clients, as
described in {{handle-server-response}}.

extensions
: A list of extensions that the client must take into consideration when
generating a ClientHello message. These are described below
({{config-extensions}}).

The `HpkeKeyConfig` structure contains the following fields:

> key_id

Reviewer: Why `key_id` instead of `config_id`? The new name sounds as if it should be possible to have a single key used by multiple configs, but what we actually want is a handle with which the server can resolve which config to use.

Author: Since the identifier corresponds primarily to the HPKE public key, and not the public name et al. (If we lifted the key_id *out* of the HpkeKeyConfig, I think config_id would be better.)

Reviewer: Perhaps another way to resolve this would be to change `key_id` to simply `id`. The `key_` prefix seems a bit redundant.

Author: We could do that, but then ClientECH would either have an inconsistently named field (key_id) or an ambiguously named field (id). I'd prefer to keep key_id.

~~~~
    struct {
       opaque config_id<0..255>
       ECHCipherSuite cipher_suite;

>        opaque config_id<0..255>;

Reviewer: Why change the wire format of the ECH extension? We might end up changing the type of `config_id` anyway, in which case the point is moot. However, for now, I think we should keep the extension parameters in the same order.

Author: Reverted!

>        opaque config_id<0..255>

Reviewer: nit: missing semicolon `;`

Author: ```suggestion
       opaque config_id<0..255>;
```

       opaque payload<1..2^16-1>;
    } ClientECH;
~~~~

config_id

> : The opaque byte identifier matching `ECHConfigContents.key_config.config_id`.

Reviewer 3: ```suggestion
: The ECHConfigContents.key_config.config_id for the chosen ECHConfig.
```

The value of the "encrypted_client_hello" extension in the ClientHelloOuter is
a `ClientECH` with the following values:

- `config_id`, the identifier corresponding to the chosen ECHConfig structure;
- `cipher_suite`, the client's chosen cipher suite;
- `enc`, as computed above; and
- `payload`, as computed above.

If optional configuration identifiers (see {{optional-configs}})) are used,
`config_id` SHOULD be set to a randomly generated eight bytes. Unless specified

> by the application using (D)TLS or externally configured on both sides,

Reviewer 3: So when is empty valid?

Author: It's always valid, I think. The server will just have to do trial decryption.

Reviewer 2: An empty identifier is just one possible value in the space.  It doesn't mean that other configurations aren't identified by non-empty identifiers.

Reviewer 3: What I'm saying here is that we have a way to indicate the identifier is optional, and that's to have it be 8 random bytes. So when would you use empty?

Author: Whenever the application wants it, I suppose? We're just recommending that these be random when optional. As Martin says, they don't need to be. Can you propose text if you think something else needs to be said here?

Reviewer 3: I think we should forbid the empty identifier.

Author: I'd like to not do anything here, especially given #379, wherein we move from a vector to a `uint`. Would that be OK?

- `payload`, as computed above.

If optional configuration identifiers (see {{optional-configs}})) are used,

> `key_id` MAY be set to a randomly generated eight bytes. Unless specified by

Author: That would align this with GREASE, which seems like a fine thing to do. (That is, in both GREASE and optional configs, the ID would be random.)

The client then modifies the "encrypted_client_hello" extension in
ClientHelloOuter as follows:

> - `key_id` is unchanged and contains the client's chosen HPKE key identifier.

Reviewer: Like we did for config_id, we could avoid re-transmitting the key_id by changing its type to a U8-prefixed string. IIRC there were upsides to not re-transmitting config_id.

Author: The upside is that the server doesn't have to check them for equality against HRR, but there's already a check for the cipher_suite field, so this seems like not a stretch.

Author: I reverted this change, too. It's now length prefixed so the HRR case is empty. We can discuss the size of the ID in a separate issue.

structure available for the server, it SHOULD send a GREASE {{?RFC8701}}
"encrypted_client_hello" extension in the first ClientHello as follows:

> - Set the `key_id` field to a random byte.

Reviewer: Whether this sticks out depends on how the real `key_id` is chosen. If chosen randomly then we're golden. However the chance of two randomly chosen key ids colliding is high.

If the client-facing server accepted ECH, it checks the second ClientHelloOuter
also contains the "encrypted_client_hello" extension. If not, it MUST abort the
handshake with a "missing_extension" alert. Otherwise, it checks that
ClientECH.cipher_suite and ClientECH.config_id are unchanged, and that

> ClientECH.enc is empty. If not, it MUST abort the handshake with an

Reviewer: I believe this change needs to be reverted as well. From {#client-hrr}:

> The client then modifies the "encrypted_client_hello" extension in ClientHelloOuter as follows:
> 
>   - **config_id is replaced with the empty string.**
>   - cipher_suite is unchanged and contains the client's chosen HPKE cipher suite.
>   - enc is replaced with the empty string.
>   - payload is replaced with the value computed above.

Author: Fixed.
